dbms/src/Storages/DeltaMerge/DeltaMergeStore_InternalBg.cpp

```diff
@@ -388,15 +430,183 @@ bool shouldCompactStableWithTooMuchDataOutOfSegmentRange(const DMContext & conte
     seg->setValidDataRatioChecked();
     return (valid_rows < total_rows * (1 - invalid_data_ratio_threshold)) || (valid_bytes < total_bytes * (1 - invalid_data_ratio_threshold));
 }
+
 } // namespace GC
 
+SegmentPtr DeltaMergeStore::gcTrySegmentMerge(const DMContextPtr & dm_context, const SegmentPtr & segment)
+{
+    auto segment_rows = segment->getEstimatedRows();
+    auto segment_bytes = segment->getEstimatedBytes();
+    if (segment_rows >= dm_context->small_segment_rows || segment_bytes >= dm_context->small_segment_bytes)
+    {
+        LOG_FMT_TRACE(
+            log,
+            "GC - Merge skipped because current segment is not small, segment={} table={}",
+            segment->simpleInfo(),
+            table_name);
+        return {};
+    }
+
+    auto segments_to_merge = getMergeableSegments(dm_context, segment);
+    if (segments_to_merge.size() < 2)
+    {
+        LOG_FMT_TRACE(
+            log,
+            "GC - Merge skipped because cannot find adjacent segments to merge, segment={} table={}",
+            segment->simpleInfo(),
+            table_name);
+        return {};
+    }
+
+    LOG_FMT_DEBUG(
```

Comment:
Is using "INFO" level better here? As this will not bring too frequent logging

Reply:
Good catch! The log level should be identical for GC merge and GC merge delta.